src/endpoint.rs

```diff
@@ -14,43 +14,62 @@
  *  limitations under the License.
  */
 
-// TODO Move endpoint.rs out of config/ into cluster/
-use crate::cluster::Endpoint;
-use std::sync::Arc;
+//! Types representing where the data is the sent.
 
-#[derive(Debug)]
-pub struct EmptyListError;
+use std::{net::SocketAddr, sync::Arc};
 
-#[derive(Debug)]
-pub struct AllEndpointsRemovedError;
+use serde::{Deserialize, Serialize};
 
-#[derive(Debug)]
-pub struct IndexOutOfRangeError;
+type EndpointMetadata = crate::metadata::MetadataView<Metadata>;
 
-/// Endpoints represents the set of all known upstream endpoints.
-#[derive(Clone, Debug, PartialEq)]
-pub struct Endpoints(Arc<Vec<Endpoint>>);
+/// A destination endpoint with any associated metadata.
+#[derive(Debug, Deserialize, Serialize, PartialEq, Clone, PartialOrd, Eq)]
+#[non_exhaustive]
+#[serde(deny_unknown_fields)]
+pub struct Endpoint {
+    pub address: SocketAddr,
+    #[serde(default)]
+    pub metadata: EndpointMetadata,
+}
 
-/// UpstreamEndpoints represents a set of endpoints.
-/// This set is guaranteed to be non-empty - any operation that would
-/// cause the set to be empty will return an error instead.
-pub struct UpstreamEndpoints {
-    /// All endpoints in the initial set - this list never
-    /// changes after initialization.
-    endpoints: Endpoints,
-    /// A view into the current subset of endpoints in the original set.
-    /// It contains indices into the initial set, to form the subset.
-    /// If unset, the initial set is the current subset.
-    subset: Option<Vec<usize>>,
+impl Endpoint {
+    /// Creates a new [`Endpoint`] with no metadata.
+    pub fn new(address: SocketAddr) -> Self {
+        Self {
+            address,
+            ..<_>::default()
+        }
+    }
+
+    /// Creates a new [`Endpoint`] with the specified `metadata`.
+    pub fn with_metadata(address: SocketAddr, metadata: impl Into<EndpointMetadata>) -> Self {
+        Self {
+            address,
+            metadata: metadata.into(),
+            ..<_>::default()
+        }
+    }
 }
 
+impl Default for Endpoint {
+    fn default() -> Self {
+        Self {
+            address: std::net::SocketAddrV6::new(std::net::Ipv6Addr::UNSPECIFIED, 0, 0, 0).into(),
+            metadata: <_>::default(),
+        }
+    }
+}
+
+/// Represents the set of all known upstream endpoints.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Endpoints(Arc<Vec<Endpoint>>);
+
 impl Endpoints {
     /// Returns an [`Endpoints`] backed by the provided list of endpoints.
-    pub fn new(endpoints: Vec<Endpoint>) -> Result<Self, EmptyListError> {
-        if endpoints.is_empty() {
-            Err(EmptyListError)
-        } else {
-            Ok(Self(Arc::new(endpoints)))
+    pub fn new(endpoints: Vec<Endpoint>) -> Option<Self> {
+        match endpoints.is_empty() {
+            true => None,
+            false => Some(Self(Arc::new(endpoints))),
         }
```

Comment:
```suggestion
        if endpoints.is_empty() {
            None
        } else {
            Some(Self(Arc::new(endpoints))),
        }
```
nit: matching on a bool feels overkill